### PR TITLE
Fix 'last' and add 'first' Spark aggregate functions

### PR DESCRIPTION
**Pull request by Yohahaha**
Use different function names to distinguish whether ignore null and
refactor 'first' and 'last' common code into one file.

Fix: #4482 #4538

### CODE REVIEW

**Comment by Yohahaha** [comment]
Hi @mbasmanova could you help take a look?

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    auto i = rows.begin();
+    while (i < rows.end()) {
+      if (!rows.isValid(i)) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:170]: use rows.applyToSelected

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    rows.applyToSelected(
+        [&selectedIndexes](vector_size_t i) { selectedIndexes.push_back(i); });
+
+    if (decoded.isConstantMapping()) {
+      if (!decoded.isNullAt(0)) {
+        rows.applyToSelected(
+            [&](vector_size_t i) { this->updateValue(i, groups[i], decoded); });
+      }
+    } else if (decoded.mayHaveNulls()) {
+      for (auto it = selectedIndexes.crbegin(); it < selectedIndexes.crend();
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:137]: Any reason not to use rows.applyToSelected ?

File: velox/functions/sparksql/coverage/Coverage.cpp [hunk 3]
```diff
@@ -31,7 +31,7 @@ int main(int argc, char** argv) {
   functions::sparksql::registerFunctions("");
 
   // Register Spark aggregate functions.
-  functions::sparksql::aggregates::registerAggregateFunctions("");
+  functions::sparksql::aggregate::registerAggregateFunctions("");
```
Comment (mbasmanova) [inline velox/functions/sparksql/coverage/Coverage.cpp:34]: This change makes this PR hard to review. Would it be possible to extract this refactor into a separate PR?

**Review by mbasmanova — COMMENTED** [review]
@Yohahaha Thank you for fixing 'last' aggregate function adding 'first' aggregate function. Let's clarify the PR title: "Fix 'last' and add 'first' Spark aggregate functions".

> Add disableTestStreaming method in AggregationTestBase.h to disable test streaming aggregation, first aggregate function is stateless, not suitable in current stream aggregation test code.

What makes this function not suitable for streaming aggregation?

**Comment by Yohahaha** [comment]
>> Add disableTestStreaming method in AggregationTestBase.h to disable test streaming aggregation, first aggregate function is stateless, not suitable in current stream aggregation test code.

> What makes this function not suitable for streaming aggregation?

PR #3965 add test streaming aggregation logic of these code https://github.com/facebookincubator/velox/pull/3965/files#diff-6be91685198b431695c5fb583b01b2787d1dcd4f05bee6203b624e8a2d7dd153R415-R444
However, above test code assume aggregate function is stateful, agg func b can resume its state from intermediate result of agg func a. But for last/first agg func, it accept a group of rows and compute a result, result has not related with input rows' order or rows which has been computed, these aggregate function are stateless.

I have less background informations about PR3965, so I just add a method a control `testStreaming`, similar to `allowInputShuffle`
https://github.com/facebookincubator/velox/blob/3bb82fe4efb9d68996a05c1946aa16ef28139bd3/velox/functions/prestosql/aggregates/tests/AggregationTestBase.h#L129-L133

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp (hunk 2 above)
Comment (Yohahaha) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:137]: `rows.applyToSelected` loop all valid rows in normal order, in first aggregate function, we need loop all valid rows in reversed order, cause first row of group is we need.

I try to add a new method of `rows.applyToSelectedReverse`, but I have no ideas about how to implement it with `bits::forEachSetBit`, so I collect all valid row index then loop with reverse order.

File: velox/functions/sparksql/coverage/Coverage.cpp (hunk 3 above)
Comment (Yohahaha) [reply velox/functions/sparksql/coverage/Coverage.cpp:34]: Got it, I will revert and resubmit in another PR.

**Comment by mbasmanova** [comment]
@Yohahaha 

PR #3965 added test logic to ensure that an aggregate function can receive raw input, produce intermediate result, get initialized from intermediate result, receive some more raw input and produce final result. This is a pattern used in streaming systems like Flink. I do not see any reason why 'last' or 'first' Spark function cannot support this pattern. Would you investigate some more to find out what's going on?

CC: @Yuhta

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:137]: I'm not sure this is correct as 'rows' contains different groups. 

Why can't we use rows.applyToSelected and implement updateValue for 'first' to keep the first value, while for 'last' to keep the last value?

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+ protected:
+  void updateValue(vector_size_t i, char* group, DecodedVector& decoded) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:98]: I expect this logic to be different for 'first' and 'last' functions.

'first' function would store the value only if currently there is no value.

'last' function would store the value unconditionally.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp (hunk 1 above)
Comment (Yohahaha) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:170]: BTW, this logic is comes from old `LastAggregate.cpp`, I think it's better than `rows.applyToSelected`

Comment (mbasmanova) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:170]: BTW, if you want to implement early return, then use rows.testSelected.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+      bool mayPushdown) override {
+    addSingleGroupRawInput(group, rows, args, mayPushdown);
+  }
+};
+
+} // namespace
+
+template <template <bool B1, bool B2, typename T> class TClass, bool ignoreNull>
+bool registerFirstLast(const std::string& name) {
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures;
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:273]: Why can't we support all types?

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+        "double",
+        "varchar"}) {
+    signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                             .argumentType(inputType)
+                             .intermediateType(inputType)
+                             .returnType(inputType)
+                             .build());
+  }
+  signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                           .typeVariable("T")
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:289]: This signature implies that we support arrays of *all* types.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp (hunk 1 above)
Comment (Yohahaha) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:170]: > What happens if we use rows.applyToSelected here?

It's fine, I just want early return like you said, I will try to use `rows.testSelected`, thanks!

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp (hunk 2 above)
Comment (Yohahaha) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:137]: > I'm not sure this is correct as 'rows' contains different groups.

I have verified this logic, it's correct.

> Why can't we use rows.applyToSelected and implement updateValue for 'first' to keep the first value, while for 'last' to keep the last value?

This way is correct in global aggregation which all rows in one group, the rows are deterministic and we only maintain one value.
But for aggregation with group by, Velox will pass all valid rows from all group in `addRawInput/addIntermediateResults`, it's hard and impossible to separate rows from different group, and we have no idea about how many groups need to process, so can not maintain a internal vector of values here.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp (hunk 4 above)
Comment (Yohahaha) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:98]: Is any way to check whether `group` is set or not?

Loop all rows and simply update group value may better than explicitly compare and set?

**Comment by Yohahaha** [comment]
> @Yohahaha
> 
> PR #3965 added test logic to ensure that an aggregate function can receive raw input, produce intermediate result, get initialized from intermediate result, receive some more raw input and produce final result. This is a pattern used in streaming systems like Flink. I do not see any reason why 'last' or 'first' Spark function cannot support this pattern. Would you investigate some more to find out what's going on?
> 
> CC: @Yuhta

Flink is a streaming process engine, operators are stateful, intermediate results or state need persisted in somewhere for , but Spark's most operators are stateless used in batch process scenario , Spark also support streaming computation which operators are stateful. https://stackoverflow.com/questions/49089866/what-is-the-difference-between-a-stateful-and-stateless-system

Actually, I am not sure Velox wants to be a streaming process engine or batch process engine, and we may have different understand for current StreamingAggregation in Velox, I think it does not equal to streaming process.

**Comment by mbasmanova** [comment]
@Yohahaha You are correct that streaming use case of Flink and StreamingAggregation operator in Velox are different things. We do use Velox for multiple engines including adhoc (Presto), batch (Spark and Presto-on-Spark) and streaming (Meta-specific version of Flink). The aggregation functions are all stateful. They maintain the state in the accumulators. Like I mentioned before, I do not see why would 'last' or 'first' aggregate functions not support Flink-like use case. Please, take the time to investigate why the test is failing.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp (hunk 6 above)
Comment (mbasmanova) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:289]: What I'm trying to say is that currently the signatures are define in such a way that first(array(row(...))) is supported, but first(row(...)) is not. This doesn't quite makes sense.

Comment (mbasmanova) [reply velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:289]: Similarly, first(array(boolean)) is supported, but first(boolean) is not.

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp [hunk 7]
```diff
@@ -36,8 +36,9 @@ DEFINE_string(
     "(e.g: --only \"min\" or --only \"sum,avg\").");
 
 int main(int argc, char** argv) {
+  std::string prefix = "spark_";
```
Comment (mbasmanova) [inline velox/exec/tests/SparkAggregationFuzzerTest.cpp:39]: Why do we need the prefix? I assume this change is unrelated. Let's extract it into a separate PR.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    rows.applyToSelected(
+        [&](vector_size_t i) { updateValue(i, groups[i], decoded); });
+  }
+
+  void addIntermediateResults(
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:146]: This can go into the base class.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    rows.testSelected(
+        [&](vector_size_t i) { return updateValue(i, group, decoded); });
+  }
+
+  void addSingleGroupIntermediateResults(
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:165]: This can go into the base class.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addSingleGroupRawInput(group, rows, args, mayPushdown);
+  }
+
+ private:
+  // If we found a valid value, set accumulator valid flags to true, then skip
+  // remaining rows in group.
+  bool updateValue(vector_size_t i, char* group, DecodedVector& decoded) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:176]: Maybe this can be simplified a bit. See https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:first_last_agg?expand=1

**Review by mbasmanova — COMMENTED** [review]
@Yohahaha Thank you for iterating on this PR. Some further comments.

**Comment by Yohahaha** [comment]
@mbasmanova Thank for your all detailed cpp language suggestions! 

I'm trying support all types for these functions and add UT for new supported types, thanks again!

**Comment by Yohahaha** [comment]
Hi @mbasmanova 
All comments has been addressed, thanks again for your nice metaprograming suggestions.

CI failed due to homebrew update failed in macOS.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+using namespace facebook::velox::aggregate;
+
+/// FirstLastAggregate returns the first or last value of |expr| for a group of
+/// rows. If |ignoreNull| is true, returns only non-null values.
+///
+/// The function is non-deterministic because its results depends on the order
+/// of the rows which may be non-deterministic after a shuffle.  This can be
+/// made deterministic by providing explicit ordering by adding order by or sort
+/// by in query.
+template <bool numeric, typename TDataType>
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:36]: nit: TDataType -> TData since T already means 'type'.

File: velox/functions/sparksql/aggregates/FirstLastAggregate.cpp [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+          case TypeKind::TIMESTAMP:
+            return std::make_unique<TClass<ignoreNull, Timestamp, true>>(
+                resultType);
+          case TypeKind::DATE:
+            return std::make_unique<TClass<ignoreNull, Date, true>>(resultType);
+          case TypeKind::VARCHAR:
+          case TypeKind::ARRAY:
+          case TypeKind::MAP:
+          case TypeKind::ROW:
+            // int8_t here is a tricky for compile, has no meaning.
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/FirstLastAggregate.cpp:323]: This comment is confusing to me. What does it mean? Consider using ComplexType instead of int8_t (see Type.h).

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp [hunk 13]
```diff
@@ -48,11 +48,19 @@ int main(int argc, char** argv) {
...
+  std::unordered_set<std::string> skipFunctions = {};
+
+  // The results of the following functions depend on the order of input
+  // rows. For some functions, the result can be transformed to a value that
+  // doesn't depend on the order of inputs. If such transformation exists, it
+  // can be specified to be used for results verification. If no transformation
+  // is specified, results are not verified.
+  std::unordered_map<std::string, std::string> orderDependentFunctions = {
+      {"last", ""},
+      {"last_ignore_null", ""},
```
[the remark this answers is not in the document]
Comment (Yohahaha) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp:60]: Second parameter in last function is a config, usually used in SQL. I think `ignoreNulls` should not appear in RowVectors as a column, so I use different function name to match this behavior. 

Downstream project such as Gluten would maintain a mapping of SparkSQL function to Velox function, if SparkSQL function specify `ignoreNulls`, Gluten should chose correct implementation.

[the remark this answers is not in the document]
Comment (Yohahaha) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp:60]: For `size` function, https://spark.apache.org/docs/3.3.1/api/sql/#size
> The function returns null for null input if spark.sql.legacy.sizeOfNull is set to false or spark.sql.ansi.enabled is set to true.

`spark.sql.legacy.sizeOfNull` here is static/session function config, no one will use this like `size(expr, legacySizeOfNull)` in plain SQL `select size(expr, config) from tmp`, and Velox provide the ability of use QueryConfig to initialize scalar functions. https://facebookincubator.github.io/velox/develop/scalar-functions.html#access-to-session-properties-and-constant-inputs

But for aggregate function `last`, Velox has no ability to pass QueryConfig-like config for initialization, and ignoreNull used in SQL `select last(expr, ignoreNull) from tmp`, it does not appear in RowVector as column, whether ignoreNull decided by different implementation.

> It would be nice to make be consistent and either remove spark.legacy-size-of-null and replace it with a different flavor of 'size' or add a similar config for first/last aggregate functions.

I agree, but seems design of scalar function and aggregate function are different.

I prefer keep current aggregate function design and provide two version of `last` implementation, Velox is a native library, let downstream project to chose which implementation is needed.

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp:60]: @Yohahaha Is the problem then that Aggregate functions do not have access to QueryConfig? Should we fix that?

[the remark this answers is not in the document]
Comment (Yohahaha) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp:60]: Let me open a discussion for this.

**Comment by Yohahaha** [comment]
linux-adapters failed with segment fault, not related with this PR.
https://app.circleci.com/pipelines/github/facebookincubator/velox/23195/workflows/02a9739c-60a0-4ff5-948c-ac9b54387444/jobs/146997
```
189: [ RUN      ] MinMaxByAggregationTest/MinMaxByGlobalByAggregationTest.randomMaxByGlobalBy/59
192/205 Test #200: velox_exec_test ................................................***Exception: SegFault 14.98 sec
```

linux-build failed with function signatures check, I will rebase main to pass it.

macos-build-macos-intel failed due to homebrew update timeout, can be ignore.

**Comment by Yohahaha** [comment]
linux-adapters failed with segment fault, not related with this PR.
https://app.circleci.com/pipelines/github/facebookincubator/velox/23195/workflows/02a9739c-60a0-4ff5-948c-ac9b54387444/jobs/146997
```
189: [ RUN      ] MinMaxByAggregationTest/MinMaxByGlobalByAggregationTest.randomMaxByGlobalBy/59
192/205 Test #200: velox_exec_test ................................................***Exception: SegFault 14.98 sec
```

linux-build failed with function signatures check, I will rebase main to pass it.

macos-build-macos-intel failed due to homebrew update timeout, can be ignore.

File: velox/docs/functions/spark/aggregate.rst [hunk 14]
```diff
@@ -11,10 +11,14 @@ General Aggregate Functions
 
     Returns the bitwise XOR of all non-null input values, or null if none.
 
-.. spark:function:: last(x) -> x
+.. spark:function:: first(x, ignoreNull) -> x
```
Comment (mbasmanova) [inline velox/docs/functions/spark/aggregate.rst:14]: This documentation doesn't match the code. We should document 2 functions: first(x) and first_ignore_null(x), no?

File: velox/functions/sparksql/aggregates/Register.cpp [hunk 15]
```diff
@@ -17,12 +17,13 @@
...
 #include "velox/functions/sparksql/aggregates/BitwiseXorAggregate.h"
-#include "velox/functions/sparksql/aggregates/LastAggregate.h"
 
 namespace facebook::velox::functions::sparksql::aggregates {
 
+extern void registerFirstLastAggregate(const std::string& prefix);
+
 void registerAggregateFunctions(const std::string& prefix) {
-  aggregates::registerLastAggregate(prefix + "last");
+  registerFirstLastAggregate(prefix);
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/Register.cpp:26]: nit: since this method registers multiple functions, perhaps, rename to registerFirstLastAggregates with 's' at the end

**Comment by Yohahaha** [comment]
Hi @mbasmanova 
All comments were fixed, CI passed.

File: velox/docs/functions/spark/aggregate.rst [hunk 16]
```diff
@@ -11,10 +11,18 @@ General Aggregate Functions
 
     Returns the bitwise XOR of all non-null input values, or null if none.
 
+.. spark:function:: first(x) -> x
+
+    Returns the first value of `x` for a group of rows, not ignore null value.
+
+.. spark:function:: first_ignore_null(x) -> x
+
+    Returns the first value of `x` for a group of rows, returns only non-null value.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/aggregate.rst:20]: nit: Returns the first non-null value ..

**Review by mbasmanova — APPROVED** [review]
@Yohahaha Thanks.
